### ISSUE TITLE
Don't force the Provider interface to use web3.js Connection

### ISSUE THREAD

**Issue by kevinheavey** [coral-xyz/anchor#2618]
Currently the Provider interface requires a Connection attribute:

```ts
export default interface Provider {
    readonly connection: Connection;
...
```
This forces anchor-bankrun to lie and make a fake Connection object which is quite nasty:
```ts
this.connection = new BankrunConnectionProxy(
    context.banksClient,
) as unknown as Connection; // uh
```
Since Anchor only uses a few of the Connection methods, we could just require a ConnectionInterface instead.

From searching through the code, it would need to expose a `.commitment` attribute and implement the following methods:

- getAccountInfo
- getAccountInfoAndContext
- getMinimumBalanceForRentExemption
- getProgramAccounts

For websocket support (which should probably be optional) we'd also need `onAccountChange` and `removeAccountChangeListener`.

**Comment by acheroncrypto** [comment]
> Since Anchor only uses a few of the Connection methods, we could just require a ConnectionInterface instead.

This would result with a breaking change every time we use a new method internally.

We'll most likely migrate to the `@solana/web3.js` v2 once it's out. It's probably best to get back at this after the migration since the new version is likely to affect the required interface.

**Comment by buffalojoec** [comment]
> We'll most likely migrate to the `@solana/web3.js` v2 once it's out. It's probably best to get back at this after the migration since the new version is likely to affect the required interface.

Agreed. `@solana/web3.js` v2 will allow you to do pretty much exactly what you're describing @kevinheavey.

The Anchor client's use of the Solana RPC JavaScript client does not have to be the specific client constructed within the main `@solana/web3.js` library.

The main library provides the "default" HTTP transport for those who want this thing to work out-of-the-box.

https://github.com/solana-labs/solana-web3.js/blob/687ea3dc6078fbb5dae7bc3391e9b7c7dce12409/packages/library/src/rpc-transport.ts#L20-L34

However, the RPC [core](https://github.com/solana-labs/solana-web3.js/tree/master/packages/rpc-core) and [transports](https://github.com/solana-labs/solana-web3.js/tree/master/packages/rpc-transport) are actually separate decoupled libraries.

So, Anchor can basically create its own HTTP transport however it desires, like so:

https://github.com/buffalojoec/solana-web3.js/blob/b702fac7294146a03e1072bfebce332784b66a60/packages/library-legacy/src/rpc.ts

...which, as you're describing, would just be an interface that satisfies this:

https://github.com/buffalojoec/solana-web3.js/blob/b702fac7294146a03e1072bfebce332784b66a60/packages/library-legacy/src/rpc.ts#L35-L42

where instead of `SolanaRpcMethods`, perhaps we have a subset of these methods, like `AnchorRpcMethods`
